src/couch/priv/couch_js/68/utf8.cpp

```diff
@@ -0,0 +1,310 @@
+{
+    uint32_t ucs4Char;
+    uint32_t minucs4Char;
+
+    // from Unicode 3.1, non-shortest form is illegal
+    static const uint32_t minucs4Table[] = {
+        0x00000080, 0x00000800, 0x0001000, 0x0020000, 0x0400000
+    };
+
+    if (utf8Length == 1)
+    {
+        ucs4Char = *utf8Buffer;
+    }
+    else
+    {
+        ucs4Char = *utf8Buffer++ & ((1<<(7-utf8Length))-1);
+        minucs4Char = minucs4Table[utf8Length-2];
+        while(--utf8Length)
+        {
+            ucs4Char = ucs4Char<<6 | (*utf8Buffer++ & 0x3F);
+        }
+        if(ucs4Char < minucs4Char || ucs4Char == 0xFFFE || ucs4Char == 0xFFFF)
+        {
+            ucs4Char = 0xFFFD;
+        }
+    }
+
+    return ucs4Char;
+}
+
+static bool
+dec_charbuf(const char *src, size_t srclen, char16_t *dst, size_t *dstlenp)
+{
+    uint32_t v;
+    size_t offset = 0;
+    size_t j;
+    size_t n;
+    size_t dstlen = *dstlenp;
+    size_t origDstlen = dstlen;
+
+    if(!dst) dstlen = origDstlen = (size_t) -1;
+
+    while(srclen)
+    {
+        v = (uint8_t) *src;
+        n = 1;
+        
+        if(v & 0x80)
+        {
+            while(v & (0x80 >> n))
+            {
+                n++;
+            }
+            
+            if(n > srclen) goto buffer_too_small;
+            if(n == 1 || n > 6) goto bad_character;
+            
+            for(j = 1; j < n; j++)
+            {
+                if((src[j] & 0xC0) != 0x80) goto bad_character;
+            }
+
+            v = dec_char((const uint8_t *) src, n);
+            if(v >= 0x10000)
+            {
+                v -= 0x10000;
+                
+                if(v > 0xFFFFF || dstlen < 2)
+                {
+                    *dstlenp = (origDstlen - dstlen);
+                    return false;
+                }
+                
+                if(dstlen < 2) goto buffer_too_small;
+
+                if(dst)
+                {
+                    *dst++ = (char16_t)((v >> 10) + 0xD800);
+                    v = (char16_t)((v & 0x3FF) + 0xDC00);
+                }
+                dstlen--;
+            }
+        }
+
+        if(!dstlen) goto buffer_too_small;
+        if(dst) *dst++ = (char16_t) v;
+
+        dstlen--;
+        offset += n;
+        src += n;
+        srclen -= n;
+    }
+
+    *dstlenp = (origDstlen - dstlen);
+    return true;
+
+bad_character:
+    *dstlenp = (origDstlen - dstlen);
+    return false;
+
+buffer_too_small:
+    *dstlenp = (origDstlen - dstlen);
+    return false;
+}
+
+JSString*
+dec_string(JSContext* cx, const char* bytes, size_t byteslen)
+{
+    JSString* str = NULL;
+    size_t charslen;
+    
+    if(!dec_charbuf(bytes, byteslen, NULL, &charslen)) return NULL;
+
+    JS::UniqueTwoByteChars chars(js_pod_malloc<char16_t>(charslen + 1));
+    if(!chars) return NULL;
+    chars.get()[charslen] = 0;
+
+    if(!dec_charbuf(bytes, byteslen, chars.get(), &charslen)) goto error;
+
+    str = JS_NewUCString(cx, std::move(chars), charslen - 1);
```

Comment:
The requirement here is the use of `js_pod_malloc` on L293, to prevent reaping of the memory when going out of the JS realm (this happens with the http bindings), combined with `std::move` here, necessary per the header file which actually deletes the ctor if you refuse to use it.

Drop either of those decisions and you get either double `free()`s or segfaults.

Reply:
Where did you see the bit about js_pod_malloc vs JS_alloc? Or was that just trial and error?

Reply:
Trial & error plus multiple places where this is done over in `js/src/shell/js.cpp`.

Reply:
@davisp FYI, if you leave off the `std::move()` here, compile fails with this informative error - probably intentional on Mozilla's part:

```
Compiling /home/jenkins/couchdb/src/couch/priv/couch_js/68/utf8.cpp
/home/jenkins/couchdb/src/couch/priv/couch_js/68/utf8.cpp: In function 'JSString* dec_string(JSContext*, const char*, size_t)':
/home/jenkins/couchdb/src/couch/priv/couch_js/68/utf8.cpp:299:49: error: use of deleted function 'mozilla::UniquePtr<T [], D>::UniquePtr(const mozilla::UniquePtr<T [], D>&) [with T = char16_t; D = JS::FreePolicy]'
  299 |     str = JS_NewUCString(cx, chars, charslen - 1);
      |                                                 ^
In file included from /usr/include/mozjs-68/mozilla/Span.h:31,
                 from /usr/include/mozjs-68/mozilla/Range.h:12,
                 from /usr/include/mozjs-68/jsapi.h:15,
                 from /home/jenkins/couchdb/src/couch/priv/couch_js/68/utf8.cpp:13:
/usr/include/mozjs-68/mozilla/UniquePtr.h:455:3: note: declared here
  455 |   UniquePtr(const UniquePtr& aOther) = delete;  // construct using std::move()!
      |   ^~~~~~~~~
In file included from /home/jenkins/couchdb/src/couch/priv/couch_js/68/utf8.cpp:13:
/usr/include/mozjs-68/jsapi.h:2355:70: note:   initializing argument 2 of 'JSString* JS_NewUCString(JSContext*, JS::UniqueTwoByteChars, size_t)'
 2355 |                                               JS::UniqueTwoByteChars chars,
      |                                               ~~~~~~~~~~~~~~~~~~~~~~~^~~~~
ERROR: compile failed while processing /home/jenkins/couchdb/src/couch: rebar_abort
```

Looks to me like they both intentionally deleted the function it could have used, _and_ put that helpful comment there to help you out.

That convinced me, finally, that `std::move()` here is necessary and sufficient.